### ISSUE TITLE
Can't take screenshots of antialiased render target

### ISSUE THREAD

**Issue by Wumpf** [ggez/ggez#751]
**Describe the bug**
Taking a screenshot of a window created with `.samples(conf::NumSamples::Eight)` (or any number of samples) fails with:
`VideoError("Memory copy error: OutOfSrcBounds { size: [1920, 1080, 1], copy_end: [1920, 1080, 0] }")`

**To Reproduce**
* use `.samples(conf::NumSamples::Eight)` on window_setup
* run `graphics::screenshot(ctx)`

**Expected behavior**
Ideal: GGEZ is able to copy the msaa buffer, img.encode does anti alias resolve for me.
Or alternatively resolve happens when taking the screenshot (might be easier to implement)

**Hardware and Software:**
 - ggez version: 0.5.1
 - OS: Windows 10
 - Graphics card: Nvidia

**Comment by icefoxen** [comment]
Belated thanks for the bug report.  Resolving MSAA is kind of a black art in ggez in general right now, it doesn't really have a consistent way of doing it and thus is buggy.  I don't know much about how to make it better unfortunately, so if you do I'd love to have some advice.  In Vulkan resolving MSAA and such is a particular step that occurs at a particular time when copying a buffer, but in OpenGL (and `gfx-rs`) it seems fuzzier, more like "this buffer has MSAA so you just have to know what magic flags to give when copying it in particular situations".

**Comment by Wumpf** [comment]
I think I dealt with this somewhere at some point... I'll have a look next weekend (or the one after next, can't promise anything :))

**Comment by icefoxen** [comment]
Sure, no problem.  My own response time isn't great either these days.  :wink:

**Comment by Wumpf** [comment]
Looked into it a bit. Ended up fixing the screenshot test & screenshots being upside down. But the actual issue is a bit tricky, since the aa resolve functionality is not implemented in gfx-rs for opengl and isn't exposed in the xplatform apis :(
https://github.com/gfx-rs/gfx/blob/147b64c4432fe16ef2c866b20f538dc5686bc59d/src/backend/gl/src/command.rs#L794

I'll poke a bit if I can do the resolve on cpu, but not sure it's worth it :/
Will put open a PR for the rest regardless
https://github.com/ggez/ggez/compare/master...Wumpf:fix-screenshots

**Comment by Wumpf** [comment]
gfx-rs AA resolve issue: https://github.com/gfx-rs/gfx/issues/2875

**Comment by icefoxen** [comment]
Oof, thank you for the update!  I didn't even think that might be something that gfx-rs did not support.  If we can't make this work transparently without gfx-rs addressing that issue (which doesn't seem super likely at this point), do you think it's possible to get a non-MSAA'd screenshot?  Or maybe at least fail with a useful error message?

**Comment by Wumpf** [comment]
Continued looking into this. So we can't even copy textures with anti-aliasing which is necessary to get the data to a buffer from where we can read it.
(we can't because the copy method does an `is_inside` check on the source texture info https://github.com/gfx-rs/gfx/blob/v0.18/src/render/src/encoder.rs#L288 which just always fails if there is multisampling https://github.com/gfx-rs/gfx/blob/v0.18/src/core/src/texture.rs#L379)

The only thing we could try is to screenshot the front buffer instead of the backbuffer. Front buffer should be implicitly resolved (screens don't know what to do with multisample data). But that's shaky terrain because then platform specific stuff is needed. Also since the current pipeline state is screenshot'ed the behavior would be arguably awkward. 

> Or maybe at least fail with a useful error message?

Yeah that's what I gonna do now :/

**Comment by PSteinhaus** [comment]
Now that antialiased canvases have been fixed in #991 we could probably also fix this.

We'd need to resolve the multi-sampled texture into a new non-multi-sampled texture (and a corresponding render target) that's created with the size of the current canvas/target-texture, by drawing that texture into it using the special resolve shader that ggez now uses as its MSAA workaround.
Then we can create a new `Image` with this texture and we're done.

Pointers on how to do this can probably be found in the implementations of `Image::draw` and `GraphicsContextGeneric::draw`.

**Comment by PSteinhaus** [comment]
With the wgpu-PR this has now been fixed.
To create a screenshot of an antialiased render target, i.e. a `ScreenImage` with `samples` > 1, you just do
```rust
screen_image.image(&ctx.gfx).encode(
    ctx,
    graphics::ImageEncodingFormat::Png,
    "/screenshot.png",
)?;
```
(taken from the `astroblasto` example)
